x-pack/plugins/observability_solution/slo/server/services/get_preview_data.ts

```diff
@@ -516,15 +516,20 @@ export class GetPreviewData {
 
   private getGroupingsFilter(options: Options, filter: estypes.QueryDslQueryContainer[]) {
     const groupingsKeys = Object.keys(options.groupings || []);
+
     if (groupingsKeys.length) {
       groupingsKeys.forEach((key) => {
         filter.push({
           term: { [key]: options.groupings?.[key] },
         });
       });
-    } else if (options.instanceId !== ALL_VALUE && options.groupBy) {
-      filter.push({
-        term: { [options.groupBy]: options.instanceId },
+    } else if (options.instanceId && options.instanceId !== ALL_VALUE && options.groupBy) {
+      const instanceIdPart = options.instanceId.split(',');
+      const groupByPart = Array.isArray(options.groupBy) ? options.groupBy : [options.groupBy];
+      groupByPart.forEach((groupBy, index) => {
+        filter.push({
+          term: { [groupBy]: instanceIdPart[index] },
+        });
```

Comment:
Hum... when the `slo.instanceId` is created, we sort the values to keep them stable. It's not guaranteed that the first value will match the first field.

https://github.com/elastic/kibana/blob/0be8295ed0028870074772c8d82093376482c4eb/x-pack/plugins/observability_solution/slo/server/assets/ingest_templates/slo_pipeline_template.ts#L75

 I wonder if we shouldn't change the ingest pipeline to ensure the first value matches first field, the second value matches the second field, and so on... To do this, we could simplify the ingest pipeline by getting rid of the painless script and replace it with:
 ```
    {
      dot_expander: {
        path: 'slo.groupings',
        field: '*',
      },
    },
    {
      set: {
        field: 'entity.id',
        value: [slo.groupBy]
          .flat()
          .map((field) => `{{{slo.groupings.${field}}}}`)
          .join(','),
      },
    },
```
This would ensure that first value in the `slo.instanceId` would map back to the first field in the group by list.

Reply:
I think that's a good approach. I like the simplification. We probably need to handle the case where groupBy is `*` but otherwise looks good to me

Reply:
I came up with this 2 processors to handle no groupings. Local testing seems to work fine without group, with 1 and many groups
```
 {
      dot_expander: {
        path: 'slo.groupings',
        field: '*',
        ignore_failure: true,
        if: 'ctx.slo.groupings != null',
      },
    },
    {
      set: {
        description: 'Generated the instanceId field based on the groupings field',
        field: 'slo.instanceId',
        value: [slo.groupBy].flat().includes(ALL_VALUE)
          ? ALL_VALUE
          : [slo.groupBy]
              .flat()
              .map((field) => `{{{slo.groupings.${field}}}}`)
              .join(','),
      },
    },
```